Long, one-line functions with multiple statements require two passes

This shows up with `BlueStyle()` but not the default `format`. I haven't narrowed down the precise settings to reproduce it.

If a one-line function definition is long enough that it exceeds the width to wrap, the function will expand to a long `function ... end` block with the contents reproduced exactly. However, a second `format` pass rewrites the contents of this function with another return statement.

This is a consequence of bad code style in the first place (the original function should probably use the long definition), so I'm not sure if this should be caught by `JuliaFormatter`. The problem is the outer `return` after the second pass.

Example:
```julia
julia> myfun = format_text("f(A_VERY_LOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOONG_NAME) = (1+1; return 2)", BlueStyle())
"function f(A_VERY_LOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOONG_NAME)\n    (1 + 1; return 2)\nend"

julia> myfun2 = format_text(myfun, BlueStyle())
"function f(A_VERY_LOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOONG_NAME)\n    return (1 + 1; return 2)\nend"

julia> @assert myfun == myfun2
ERROR: AssertionError: myfun == myfun2
Stacktrace:
 [1] top-level scope at REPL[33]:1```

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "JuliaFormatter"
 uuid = "98e50ef6-434e-11e9-1051-2b60c6c9e899"
 authors = ["Dominique Luna <dluna132@gmail.com>"]
-version = "0.20.3"
+version = "0.20.4"
 
 [deps]
 CSTParser = "00ebfdb7-1f24-5e51-bd34-a7502290713f"

--- a/src/passes.jl
+++ b/src/passes.jl
@@ -369,6 +369,21 @@ function prepend_return!(fst::FST, s::State)
         # fst[end-1] is a newline
         return
     end
+    # fix #426
+    # don't add return if the last node is a throw call. throw is a built-in function
+    # that shouldn't be overwritten for over purposes so this should be fine.
+    ln.typ === Call && ln[1].typ === IDENTIFIER && ln[1].val == "throw" && return
+
+    # check to see if the last node already has a return
+    found_return = false
+    f = (fst::FST, ::State) -> begin
+        if fst.typ === Return
+            found_return = true
+        end
+        return
+    end
+    walk(f, ln, s)
+    found_return && return
 
     ret = FST(Return, fst.indent)
     kw = FST(KEYWORD, -1, fst[end].startline, fst[end].endline, "return")

--- a/test/options.jl
+++ b/test/options.jl
@@ -363,6 +363,28 @@
             end
         end"""
         @test fmt(str, 4, 92, always_use_return = true) == str
+
+        @testset "426" begin
+            # throw function heuristic
+            str = """
+            function f(x)
+                x == 1 && return "a"
+                x == 2 && return "b"
+                throw(ArgumentError("x must be 1 or 2"))
+            end
+            """
+            @test fmt(str, always_use_return = true) == str
+        end
+
+        @testset "507" begin
+            # detect return
+            str = """
+            function f()
+                (1 + 1; return 2)
+            end
+            """
+            @test fmt(str, always_use_return = true) == str
+        end
     end
 
     @testset "whitespace in keyword arguments" begin

